### PR TITLE
feat: efficient current Dock item detection and improved window management

### PR DESCRIPTION
**Pull request by ShlomoCode**
## Describe your changes

- Get current dock item element by macOS Dock's detection (`kAXSelectedChildrenChangedNotification` and `kAXSelectedChildrenAttribute`) instead of calculating mouse position.
- Get rid of listening for mouse events, reducing CPU usage to 0% when not interacting\hovering with the Dock.
- Store windows in cache (SpaceWindowCacheManager.swift) by PID instead of BundleID, to deal with apps like scrcpy that not having a BundleID.
- Adds additional logic for better window state management. 
- Addresses window switcher update inconsistencies. 
- Introduces window fade out animations and configuration options. 

Ported from https://github.com/ejbills/DockDoor/pull/214.

## Related issue number(s) and link(s)
closes https://github.com/ejbills/DockDoor/issues/190
closes https://github.com/ejbills/DockDoor/issues/19
closes https://github.com/ejbills/DockDoor/issues/243
closes https://github.com/ejbills/DockDoor/issues/30
closes https://github.com/ejbills/DockDoor/issues/54
closes https://github.com/ejbills/DockDoor/issues/212
closes https://github.com/ejbills/DockDoor/issues/277

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If this change affects core functionality, I have added a description highlighting the changes
- [x] I have followed conventional commit guidelines

## Core Functionality Changes
If this PR modifies core functionality, please provide a brief description of the changes and their impact below:

Fixes some bugs, and significantly improves performance.

### CODE REVIEW

**Comment by ShlomoCode** [comment]
The Dock functionality works, the window changer is still problematic.
Main points remaining:
1. https://github.com/ejbills/DockDoor/pull/254/files#diff-017b6b046f3844ad59afc9f7acd992ad75e9b77ec27f24b4186d13130fc73f20R85
2. The windows switcher is very slow to open
3. Maybe we should get the windows using kAXWindowsAttribute

Also, the code needs polishing, and better documentation of the changes in PR.

File: DockDoor/Utilities/KeybindHelper.swift [hunk 1]
```diff
@@ -110,18 +110,14 @@ class KeybindHelper {
 
     private func showHoverWindow() {
         Task { [weak self] in
-            do {
+            guard let self else { return }
+            await MainActor.run { [weak self] in
                 guard let self else { return }
-                let windows = try await WindowUtil.activeWindows(for: "")
```
Comment (ejbills) [inline DockDoor/Utilities/KeybindHelper.swift:115]: The activeWindows call to the empty string also serves to update the space cache, I don't think we need to change this part.

File: DockDoor/Utilities/SpaceWindowCacheManager.swift
Comment (ejbills) [inline DockDoor/Utilities/SpaceWindowCacheManager.swift:1]: What all exactly changes here? Are you removing all the bundle id helper functions as you are instead storing the process identifier?

Comment (ShlomoCode) [reply DockDoor/Utilities/SpaceWindowCacheManager.swift:1]: Yes, to deal with Scrcpy not having a bundle id

File: DockDoor/Utilities/KeybindHelper.swift (hunk 1 above)
Comment (ShlomoCode) [reply DockDoor/Utilities/KeybindHelper.swift:120]: What cache update do you mean?

**Comment by ShlomoCode** [comment]
@ejbills I'd appreciate help with https://github.com/ejbills/DockDoor/pull/254/files#diff-017b6b046f3844ad59afc9f7acd992ad75e9b77ec27f24b4186d13130fc73f20R85

**Comment by ejbills** [comment]
> @ejbills I'd appreciate help with https://github.com/ejbills/DockDoor/pull/254/files#diff-017b6b046f3844ad59afc9f7acd992ad75e9b77ec27f24b4186d13130fc73f20R85

Took a look, I think we can just store the previous app name string and only allow the hide action when it differs. I believe that is what happened previously.

**Comment by ShlomoCode** [comment]
You can push. You'll probably want to use DockObserver's `lastAppUnderMouse`.

**Comment by ShlomoCode** [comment]
Looks better now, But it seems like quick mouse movement doesn't trigger SwiftUI's `.onHover` :thinking:

File: DockDoor/Utilities/KeybindHelper.swift (hunk 1 above)
Comment (ejbills) [reply DockDoor/Utilities/KeybindHelper.swift:120]: For tracking windows in all spaces, the call to activeWindows is needed here.

**Comment by ejbills** [comment]
> You can push. You'll probably want to use DockObserver's `lastAppUnderMouse`.

I just sat down to work on this, did you already do it? The link goes to a different part of the code now, I can't find the TODO.

**Comment by ejbills** [comment]
> I push some changes, Looks better now, But it seems like quick mouse movement doesn't trigger SwiftUI's `.onHover` 🤔

I think it's all working on my end, what isn't working for you?

**Comment by ShlomoCode** [comment]
> > You can push. You'll probably want to use DockObserver's `lastAppUnderMouse`.
> 
> I just sat down to work on this, did you already do it? The link goes to a different part of the code now, I can't find the TODO.

Yes i did it

**Comment by ShlomoCode** [comment]
> > I push some changes, Looks better now, But it seems like quick mouse movement doesn't trigger SwiftUI's `.onHover` 🤔
> 
> I think it's all working on my end, what isn't working for you?

1. Tap on an app icon
2. A preview will appear
3. **Quickly** move the mouse cursor up through the preview
4. The preview is not hidden. If you move the mouse slowly, the preview does close properly

https://github.com/user-attachments/assets/d081bda1-331d-497d-8a3e-5c6c8fcd0513

File: DockDoor/Utilities/KeybindHelper.swift (hunk 1 above)
Comment (ShlomoCode) [reply DockDoor/Utilities/KeybindHelper.swift:120]: Didn't calling `activeWindows` with an empty string simply return all that was in the cache?
https://github.com/ejbills/DockDoor/blob/31f8c1b2c8f478e9f85d1dea692f98db715e56b9/DockDoor/Utilities/WindowUtil.swift#L409-L411

Comment (ejbills) [reply DockDoor/Utilities/KeybindHelper.swift:120]: It did indeed return `getAllWindows`, but not till after it had made a few calls to `fetchWindowInfo` which were called before the return (even with an empty string). And the `fetchWindowInfo` func served to also update the cache [here.](https://github.com/ejbills/DockDoor/blob/31f8c1b2c8f478e9f85d1dea692f98db715e56b9/DockDoor/Utilities/WindowUtil.swift#L473)

**Comment by ejbills** [comment]
> > > I push some changes, Looks better now, But it seems like quick mouse movement doesn't trigger SwiftUI's `.onHover` 🤔
> > 
> > 
> > I think it's all working on my end, what isn't working for you?
> 
>     1. Tap on an app icon
> 
>     2. A preview will appear
> 
>     3. **Quickly** move the mouse cursor up through the preview
> 
>     4. The preview is not hidden. If you move the mouse slowly, the preview does close properly
> 
> 
> CleanShot.2024-08-22.at.06.58.25.mp4

I can't get this to happen, but if I use it while I have an app fullscreen and I have the full size window hover action, the app seems to freak out. Can you reproduce?

File: DockDoor/Utilities/KeybindHelper.swift (hunk 1 above)
Comment (ShlomoCode) [reply DockDoor/Utilities/KeybindHelper.swift:120]: Okay, I got it. I don't think it should be done implicitly, I'll add a separate "cache refresh" function and call it explicitly.

Comment (ejbills) [reply DockDoor/Utilities/KeybindHelper.swift:120]: Fair enough

**Comment by ShlomoCode** [comment]
> but if I use it while I have an app fullscreen and I have the full size window hover action, the app seems to freak out. Can you reproduce?

No...
https://github.com/user-attachments/assets/fd45fef7-75e1-467d-8748-7fd6da5a9528

**Comment by ejbills** [comment]
> > but if I use it while I have an app fullscreen and I have the full size window hover action, the app seems to freak out. Can you reproduce?
> 
> No...
> 
>  Screen.Recording.2024-08-22.at.7.24.32-trimmed-converted.mov

I mean using DockDoor while inside the fullscreen app

**Comment by ShlomoCode** [comment]
Oh yes it fails to show the fullsize preview but it doesn't work for me with v1.1.5 either

**Comment by ejbills** [comment]
@ShlomoCode what is remaining for this PR? Anywhere I can help out?

**Comment by ShlomoCode** [comment]
[this](https://github.com/ejbills/DockDoor/pull/254#issuecomment-2303628342) not resolved for me yet

**Comment by ejbills** [comment]
@ShlomoCode I think I fixed everything, is it good to go now? Mind helping me by testing it?

File: DockDoor/Utilities/KeybindHelper.swift (hunk 1 above)
Comment (ejbills) [reply DockDoor/Utilities/KeybindHelper.swift]: I added in the explicit call to updating the methods as discussed, so this discussion is resolved.

**Comment by ejbills** [comment]
@ShlomoCode Also just pushed a really nice fade out effect that makes DockDoor feel really nice to use. Check it out, benefits of the changes I made with the hover detection.

**Comment by ShlomoCode** [comment]
The fade effect is nice, but is:
1. Too long (Maybe make it user-configurable)
2. If you move the mouse diagonally out, there is no fade effect:
https://github.com/user-attachments/assets/66bf9600-85b6-4a90-ae22-54c12c9592d7

File: DockDoor/Components/MouseTrackingView.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    private func setupTrackingArea() {
+        let options: NSTrackingArea.Options = [.mouseEnteredAndExited, .activeAlways, .inVisibleRect]
+        let trackingArea = NSTrackingArea(rect: bounds, options: options, owner: self, userInfo: nil)
+        addTrackingArea(trackingArea)
+    }
+
+    override func mouseExited(with event: NSEvent) {
+        DispatchQueue.main.async {
+            SharedPreviewWindowCoordinator.shared.hideWindow()
```
Comment (ShlomoCode) [inline DockDoor/Components/MouseTrackingView.swift:89]: You missed the check
```
currentAppUnderMouse == nil || currentAppUnderMouse != lastAppUnderMouse
```
This means that if you stand on an app icon, go to a preview, and return to the icon, the preview will close instead of staying open.

**Review by ShlomoCode — COMMENTED** [review]
.

**Review by ShlomoCode — COMMENTED** [review]
--

File: DockDoor/Components/MouseTrackingView.swift (hunk 2 above)
Comment (ejbills) [reply DockDoor/Components/MouseTrackingView.swift:89]: I found that it didn't require this check anymore, when I hover back on the icon it re-appears properly. Adding the check back is trivial, I will iron it out today.

**Comment by ejbills** [comment]
> The fade effect is nice, but is:
> 
> 1. Too long (Maybe make it user-configurable)
> 2. If you move the mouse diagonally out, there is no fade effect:
> 
>  CleanShot.2024-08-31.at.21.04.16.mp4



Not sure what your video is showing in regards to point 2, you can change the delay to the next application in the settings. It appears you are hovering over Xcode, which of course will hide the window for the other application. However, in your video I do see that the opacity of the view does not reset for you when you hover back onto the app. I will look into this.

**Comment by ShlomoCode** [comment]
Yes the video is not good, CleanShot X seems to be having trouble capturing the preview for some reason
What about the length of the effect? He seems too annoying to me right now.

**Comment by ejbills** [comment]
> What about the length of the effect? He seems too annoying to me right now.

I can tie that into a Default easily. 

> CleanShot X seems to be having trouble capturing the preview for some reason

Oh ok, so it is resetting properly?

**Comment by ShlomoCode** [comment]
https://github.com/user-attachments/assets/adad5d11-29dd-4d4f-8e1b-b7c72bc87245


Note that the preview disappears with no effect when you move the mouse over the icon without a preview

**Comment by ejbills** [comment]
> CleanShot.2024-08-31.at.22.11.43.mp4
> 
> Note that the preview disappears with no effect when you move the mouse over the icon without a preview

Fixed

**Comment by ejbills** [comment]
@ShlomoCode try out the new commits, I think it's good to go. I just need to test it on some external monitors. I will merge it later today if everything looks good to you.

**Comment by ShlomoCode** [comment]
> This is intended behavior

This is inconsistent behavior. If the preview disappears with an effect, the effect should always be there

**Comment by ejbills** [comment]
> This is inconsistent behavior. If the preview disappears with an effect, the effect should always be there

I understand the confusion, but let me clarify the two distinct behaviors:

1. Mouse interaction with hover window: When you interact with the hover window and then move the mouse away, there's a fade effect as the window disappears.

2. Changing hovered dock element: When you move quickly from one dock element to another, the hover window hides instantly.

These are intentionally different behaviors designed for different user interactions. The fade effect provides a smooth transition when you're done viewing a window, while the instant hiding allows for quick navigation between dock elements without visual lag.

File: DockDoor/Views/Settings/MainSettingsView.swift [hunk 3]
```diff
@@ -60,7 +60,7 @@ struct MainSettingsView: View {
             Divider()
 
             HStack {
-                Text("Hover Window Open Delay")
+                Text("Preview Window Open Delay")
```
Comment (ejbills) [inline DockDoor/Views/Settings/MainSettingsView.swift:63]: The verbiage "preview" is used for the preview hover actions. Hover window is used throughout the app and codebase to describe what it is, the hover window. Why is this changed?

File: DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift [hunk 4]
```diff
@@ -64,7 +64,7 @@ final class SharedPreviewWindowCoordinator: NSWindow {
         isMovableByWindowBackground = false
         collectionBehavior = [.canJoinAllSpaces]
         backgroundColor = .clear
-        hasShadow = false
+        hasShadow = true
```
Comment (ShlomoCode) [inline DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift:67]: This has been changed at https://github.com/ShlomoCode/DockDoor/commit/9c27ffa29f1e88bb57240b89943d0cd07178966e. Is this a deliberate change?

**Review by ShlomoCode — COMMENTED** [review]
.

**Review by ShlomoCode — COMMENTED** [review]
.

**Review by ShlomoCode — COMMENTED** [review]
.

File: DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift (hunk 4 above)
Comment (ejbills) [reply DockDoor/Views/Hover Window/SharedPreviewWindowCoordinator.swift:67]: Yes

File: DockDoor/Views/Settings/MainSettingsView.swift (hunk 3 above)
Comment (ShlomoCode) [reply DockDoor/Views/Settings/MainSettingsView.swift:63]: I think for the users "preview" is much clearer than "hover", as it reflects the essence of the window and not its hierarchy, like the window switcher is a "switcher" and not an "overlay"

**Comment by ShlomoCode** [comment]
There are still cases where the preview stays open incorrectly: https://share.cleanshot.com/rRnsHrt99TqdSx2zmCXm

It happens to me from time to time even with normal use.

**Comment by ejbills** [comment]
> There are still cases where the preview stays open incorrectly: https://share.cleanshot.com/rRnsHrt99TqdSx2zmCXm
> 
> 
> 
> It happens to me from time to time even with normal use. I don't think it can be merged yet...

All the issues related to entering the hover preview have been resolved. The issue you are showcasing here is hovering over a dock element, and then leaving the dock area without interacting with the hover preview at all, therefore the mouseExited does not execute. 

At this point, I don't think we can account for this beyond tracking the mouse outside of the dock area. Solving this issue will involve coding another handler which can catch the case the mouse is far away from the dock, and a preview is showing. Thoughts?

**Comment by ShlomoCode** [comment]
> All the issues related to entering the hover preview have been resolved. The issue you are showcasing here is hovering over a dock element, and then leaving the dock area without interacting with the hover preview at all, therefore the mouseExited does not execute.
> 

Good catch! But `handleSelectedDockItemChangedNotification` in `DockObserver.swift` is supposed to handle such a case, so that's probably where the bug lies (maybe it's related to `hoverProcessingTask?.cancel()`). Can you try to check it? I'm afraid I don't have time for that right now.

**Comment by ShlomoCode** [comment]
@ejbills If I set the Fade Out Duration to 0.1 or 0 there is a strange behavior

https://github.com/user-attachments/assets/66a5bdea-8b74-48f3-83f2-4a9daa3ffc72

**Comment by ejbills** [comment]
> `handleSelectedDockItemChangedNotification`



> > All the issues related to entering the hover preview have been resolved. The issue you are showcasing here is hovering over a dock element, and then leaving the dock area without interacting with the hover preview at all, therefore the mouseExited does not execute.
> 
> Good catch! But `handleSelectedDockItemChangedNotification` in `DockObserver.swift` is supposed to handle such a case, so that's probably where the bug lies (maybe it's related to `hoverProcessingTask?.cancel()`). Can you try to check it? I'm afraid I don't have time for that right now.

This bug is present in prod, the issue is exacerbated because the global mouse listener no longer calls the hide window func. The issue isn't the task processing - not exactly sure what the root cause is.

**Comment by ejbills** [comment]
> @ejbills If I set the Fade Out Duration to 0.1 or 0 there is a strange behavior
> CleanShot.2024-09-01.at.01.45.35.mp4

I can reproduce, I'll fix it tomorrow or later.

**Comment by ejbills** [comment]
@ShlomoCode NSTrackingArea seems to have some reliability issues when the mouse quickly crosses its area. My next course of action is another global mouse listener, but it only lives during the lifespan of the preview window (so there wouldn't be constant CPU usage still). But, I'm not sure that's the best solution. 

I have tested everything else, this is the last bug for this PR. Need some help on this issue, I'm not too sure where to go from here.

**Comment by ejbills** [comment]
@ShlomoCode we can get around this by always hiding any window before presenting a new one in the DockObserver, see commit https://github.com/ejbills/DockDoor/pull/254/commits/bbec88ea3c5883ece840dd9a442728f0d972caa9.

This appears to fix the issue, but want your thoughts on it.

**Comment by ShlomoCode** [comment]
> @ShlomoCode It appears moving the mouse quickly makes the `processSelectedDockItemChanged` method execute twice

Maybe we should implement a debounce mechanism?

**Comment by ShlomoCode** [comment]
The TEMPORARY PATCH commit did not solve the previews staying open incorrectly for me. It seems to moderate but not completely resolve it.

**Comment by ejbills** [comment]
> The TEMPORARY PATCH commit did not solve the previews staying open incorrectly for me. It seems to moderate but not completely resolve it.

It is only meant to fix hovering back onto an app, from the apps preview. That was causing it to stay open when the delay was 0.1, or 0. It does not resolve the window staying open when you don't interact with the window at all.

**Comment by ejbills** [comment]
@ShlomoCode I have solved all issues.
- Hovering back onto an application does not hide the window.
- Quickly passing over a view no longer incorrectly displays a lingering window.
- Lingering windows get cleaned up with a temporary mouse polling timer, resolving issues where the mouse does not enter the window preview at all.

I have played around with this, and I think it irons out all the issues we were dealing with. Please give it a test, and let me know what you think. I left some comments in the WindowDismissalContainer to explain this logic.

**Comment by ShlomoCode** [comment]
@ejbills Sounds good! I'll give it a spin and come back

**Comment by ShlomoCode** [comment]
It seems that when there are several open apps adjacent to each other in the Dock, the issue still exists


https://github.com/user-attachments/assets/c91a43b9-14ab-4f5c-9a72-f400f9141dc1

**Comment by ejbills** [comment]
> It seems that when there are several open apps adjacent to each other in the Dock, the issue still exists

This is the case of point 3 on this comment: https://github.com/ejbills/DockDoor/pull/254#issuecomment-2325385556
> Lingering windows get cleaned up with a temporary mouse polling timer, resolving issues where the mouse does not enter the window preview at all.

The dead zone for cleaning up lingering windows is 800 px. If you move your mouse away while a window is lingering, it will hide after the distance is > 800. I actually think this is too large of an area, what do you think is a reasonable area for this? (You can play around with it on line 59 of `WindowDismissalContainer.swift`)

**Comment by ejbills** [comment]
I have tested this all day with no issues, I will be merging. Again, thanks a ton @ShlomoCode, this is a huge and very beneficial PR for DockDoor. Great work.
